executor/compiler.go

```diff
@@ -46,17 +49,144 @@ func (c *Compiler) Compile(goCtx goctx.Context, stmtNode ast.StmtNode) (*ExecStm
 		return nil, errors.Trace(err)
 	}
 
+	CountStmtNode(stmtNode, c.Ctx.GetSessionVars().InRestrictedSQL)
+	isExpensive := logExpensiveQuery(stmtNode, finalPlan)
+
 	return &ExecStmt{
 		InfoSchema: infoSchema,
 		Plan:       finalPlan,
-		Expensive:  stmtCount(stmtNode, finalPlan, c.Ctx.GetSessionVars().InRestrictedSQL),
+		Expensive:  isExpensive,
 		Cacheable:  plan.Cacheable(stmtNode),
 		Text:       stmtNode.Text(),
 		StmtNode:   stmtNode,
 		Ctx:        c.Ctx,
 	}, nil
 }
 
+func logExpensiveQuery(stmtNode ast.StmtNode, finalPlan plan.Plan) (expensive bool) {
+	expensive = isExpensiveQuery(finalPlan)
+	if !expensive {
+		return
+	}
+
+	const logSQLLen = 1024
+	sql := stmtNode.Text()
+	if len(sql) > logSQLLen {
+		sql = fmt.Sprintf("%s len(%d)", sql[:logSQLLen], len(sql))
+	}
+	log.Warnf("[EXPENSIVE_QUERY] %s", sql)
+	return
+}
+
+func isExpensiveQuery(p plan.Plan) bool {
+	switch x := p.(type) {
+	case plan.PhysicalPlan:
+		return isPhysicalPlanExpensive(x)
+	case *plan.Execute:
+		return isExpensiveQuery(x.Plan)
+	case *plan.Insert:
+		if x.SelectPlan != nil {
+			return isPhysicalPlanExpensive(x.SelectPlan)
+		}
+	case *plan.Delete:
+		if x.SelectPlan != nil {
+			return isPhysicalPlanExpensive(x.SelectPlan)
+		}
+	case *plan.Update:
+		if x.SelectPlan != nil {
+			return isPhysicalPlanExpensive(x.SelectPlan)
+		}
+	}
+	return false
+}
+
+func isPhysicalPlanExpensive(p plan.PhysicalPlan) bool {
+	const expensiveRowThreshold = 10000
+	if p.StatsInfo().Count() > expensiveRowThreshold {
+		return true
+	}
+
+	for _, child := range p.Children() {
+		if isPhysicalPlanExpensive(child) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// CountStmtNode records the number of statements with the same type.
+func CountStmtNode(stmtNode ast.StmtNode, inRestrictedSQL bool) {
+	if inRestrictedSQL {
+		return
+	}
+	metrics.StmtNodeCounter.WithLabelValues(GetStmtLabel(stmtNode)).Inc()
+}
+
+// GetStmtLabel generates a label for a statement.
+func GetStmtLabel(stmtNode ast.StmtNode) string {
+	switch x := stmtNode.(type) {
+	case *ast.AlterTableStmt:
+		return "AlterTable"
+	case *ast.AnalyzeTableStmt:
+		return "AnalyzeTable"
+	case *ast.BeginStmt:
+		return "Begin"
+	case *ast.CommitStmt:
+		return "Commit"
+	case *ast.CreateDatabaseStmt:
+		return "CreateDatabase"
+	case *ast.CreateIndexStmt:
+		return "CreateIndex"
+	case *ast.CreateTableStmt:
+		return "CreateTable"
+	case *ast.CreateUserStmt:
+		return "CreateUser"
+	case *ast.DeleteStmt:
+		return "Delete"
+	case *ast.DropDatabaseStmt:
+		return "DropDatabase"
+	case *ast.DropIndexStmt:
+		return "DropIndex"
+	case *ast.DropTableStmt:
+		return "DropTable"
+	case *ast.ExplainStmt:
+		return "Explain"
+	case *ast.InsertStmt:
+		if x.IsReplace {
```

Comment:
I'm not sure this works, it seems Replace never show in our prometheus?

Reply:
It works, I've added a test case in **server/server_test.go**

Reply:
ok